### DISCUSSION TITLE
Images for bots

### DISCUSSION THREAD

**Discussion by tejaspatel2014** [vercel/next.js#37674] [Help]
The cached pages for google Bots do not load images unless the "Priority" prop is used. (which makes sense due to lazy loading )

You can see this behavior for demo.vercel.store (Added screenshot below for reference). The images for the first product grid would be loaded as it has priority true. However, for the rest of the page, images are not getting loaded. Apparently loading images for google-bot is helpful for SEO (Not sure what is the general opinion in the community about this as well.). 

So my question is there any known or recommended approach here. 

![screencapture-webcache-googleusercontent-search-2022-06-14-14_29_51](https://user-images.githubusercontent.com/6571086/173493636-5fa3212e-557f-4f71-9d1b-a6f0ee4ad177.png)

**Comment by icyJoseph** [comment]
My 2 cents...

## A solution without even thinking of implications

> This is almost like a bull in a glass shop

I am not sure about best practices, but if it was Friday before holiday and my Product Owner came with this issue, what I'd do is create a `_middleware` file, and use the automatic bot recognition from `request.ua.isBot` value, if `true`, rewrite the request like so:

```ts
// also probably good idea to remove `isBot` in other cases, to prevent people from toying with it
request.nextUrl.searchParams.set('isBot', '1');
return NextResponse.rewrite(request.nextUrl);
```

And then in the application look for `isBot` query, from `asPath` and load the images your are interested in with priority.

```tsx
const Home: NextPage = () => {
  const { asPath, isReady } = useRouter();
  const [isBot, setIsBot] = useState(false);

  useEffect(() => {
   if(!isReady) return;

    const isBot =
      new URLSearchParams(asPath.split("?")[1]).get("isBot") === "1";
    setIsBot(isBot);
  }, [isReady, asPath]);

  return (
    <div>
      <div style={{ height: "200vh" }}>Above the fold</div>
      <Image
        src="https://picsum.photos/200/300"
        priority={isBot}
        width="200"
        height="300"
        alt="below the fold for sure"
      />
    </div>
  );
};
```

Of course the trade-off being, that the bot will perceive the page being slow and loading more resources than necessary for FCP, and other metrics, but I'd only find that out when I come back from holiday.

## The rabbit hole

There's also the case of bots not running JS, aka JS disabled situations, where Next.js uses the native lazy loading strategy, I wonder if that also causes the issue you document. In such cases, it is the browser that decides with some heuristics when to load the image. 

https://web.dev/browser-level-image-lazy-loading/ this document which I think is the genesis of the next/image component, only talks about ads and such, and the idea of printing the page, but nothing about SEO downturns. 

If anything, there seems to be tons of pages talking about lazy loading as a net positive for SEO because of all the metrics that are improved, and sort of expect the crawler/bot to trigger the loading of the image by scrolling by (at least that's what I understand).

**Reply by tejaspatel2014** [reply]
Thanks for your input. Just adding more info. No JS Script bot or browsers are not a problem as next/image component has a fallback for no Javascript and it loads all images.

**Reply by icyJoseph** [reply]
Yes, I did mention JS disabled situations. The point I made is that in that case, it uses the native lazy loading. And then, do we see the same behaviour you are showing. The native lazy loading has its own rules and heuristics, so perhaps if the images are further down, enough, then they won't load at all.

**Reply by tejaspatel2014** [reply]
It loads successfully even if it's further down.

**Comment by ermakov95** [comment]
I have the same problem. It's strange that Next.js there is no built-in solution. The isBot option looks fragile.

Another discussion of this problem: https://github.com/vercel/next.js/discussions/39801

I figured out a little bit how it works. html comes like this:
![image](https://user-images.githubusercontent.com/89069249/226325217-c6a0687c-aee1-4e45-8937-526bb32823dd.png)
Initially there is an img tag, but with a strange src and there is a noscript tag with a normal img inside.

![image](https://user-images.githubusercontent.com/89069249/226322414-20129eaa-8a34-4fcb-97ed-c659e0b4ee03.png)
Then Next.js deletes the contents of the noscript tag and fills in the src in img and the picture is loaded and displayed.

However, in webcache.googleusercontent.com this does not happen and the html Image is output in its original form, so the image is not displayed.
![image](https://user-images.githubusercontent.com/89069249/226323177-6c194867-93b1-42bd-b0c0-54da6c8bd8a2.png)

If on webcache.googleusercontent.com turn off JS, then the images will be displayed from noscript.

"priority" solves this problem, but removes lazy-loading

**Reply by philwolstenholme** [reply]
@ermakov95 What version of Next are you using? Version 13 has an upgraded `next/image` which uses the browser's built-in native lazyloading behaviour. It'll probably help you here. You can update to version 13 without having to use all the new `app` directory/router functionality.

https://nextjs.org/docs/messages/next-image-upgrade-to-13
https://nextjs.org/blog/next-13#nextimage

**Reply by ermakov95** [reply]
I use 12.2. Thanks for the advice!
